### PR TITLE
Fixed and improved handling max_cons_per_ip.

### PR DESCRIPTION
**Pull request by sirkonst**
Fixed trace:

```
2012-09-11 10:25:36,592 - ERROR - []@127.0.0.1:59672 Connection tracking failed:
---------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.6/multiprocessing/managers.py", line 216, in serve_client
    obj, exposed, gettypeid = id_to_obj[ident]
KeyError: '7f0ac0c02228'
---------------------------------------------------------------------------
2012-09-11 10:25:36,593 - ERROR - []@127.0.0.1:59672 unhandled exception in instance <ftpcloudfs.monkeypatching.MyFTPHandler object at 0x2997510>
Traceback (most recent call last):
  File "/usr/lib/pymodules/python2.6/pyftpdlib/ftpserver.py", line 3810, in handle_accept
    handler.handle()
  File "/usr/lib/pymodules/python2.6/ftpcloudfs/monkeypatching.py", line 69, in handle
    self.logline("Connection track: %s -> %s" % (self.remote_ip, count+1))
UnboundLocalError: local variable 'count' referenced before assignment
```

And moved handling max_cons_per_ip from FTPServer to MyFTPHandler. Without it sometime FTP can accept more then max_cons_per_ip (tested by telnet).

### CODE REVIEW

**Comment by reidrac** [comment]
Lot of stuff going on here :)

There's actually a bug in the logline reporting that the lock failed (tha's the UnboundLocalError).

The reason why we're using max_cons_per_ip on ftpserver.FTPServer.max_cons_per_ip because that functionality is supported by pyftpdlib as long as you have just 1 worker.

The shared ip part using the multiprocess manager is to add max_cons_per_ip functionality when there are several workers. So effectively we're using only one variable. I can't see the advantage of adding that variable to the handler too. Please elaborate.

Secondly the RemoteError we're catching in handle is there to catch a possible self.shared_lock.acquire() fail. So having the self.shared_lock.release() in finally it's not important.

It looks like the KeyError is similar to the race condition we fixed in sha:daa9210857d333e6e87c2180cd57139d73f1e313 ; but I can't tell what's going on. Your patch improves it because you use a temp variable (count). That's a good idea, but I doesn't explain why the KeyError happened, does it?

Summing up:
1. I can't see the need of moving max_cons_per_ip
2. The moving the release lock into finally looks good!
3. Using a temp variable (count) looks good, it will definitely prevent race conditions as the one fixed in the aforementioned change.

Unfortunately none of these changes addresses the KeyError you posted in the pull request. I can't accept the patch without a test that can reproduce THAT issue.

I'd love a clean patch for 2 (finally) and 3 (prevent a race condition while checking shared_ip_map out of the lock).

In the future, please submit a bug report first so we can discuss it before writing any code!

**Comment by sirkonst** [comment]
About max_cons_per_ip:
1. Run ftp-cloudfs with 2 workers and max_cons_per_ip=1.
2. Open one connect: $ telnet 127.0.0.1 21
3. Make several attempts to open a new connect in other terminal

...and you see that you can open more then 2 connect.

**Comment by reidrac** [comment]
OK, I'll check it. Thanks!

To start with I'm going to move the release lock to a finally and use the aux variable to prevent the race condition when reading the shared dict.

**Comment by sirkonst** [comment]
Ok, thanks too :-)

**Comment by reidrac** [comment]
Done: c7b309f169e39261b8768020c103c77ca77cd2a1

I can't reproduce it :(

Configuration:

```
[ftpcloudfs]
workers = 2
max-cons-per-ip = 1
memcache = 127.0.0.1:1121
```

Command line:

```
bin/ftpcloudfs -a https://auth.storage.../v1.0 --syslog -v
```

First telnet (OK):

```
$ telnet 127.0.0.1 2021
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
220 Rackspace Cloud Files 0.17.1 using pyftpdlib 0.6.0 ready.
```

Second telnet (denied):

```
$ telnet 127.0.0.1 2021
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
220 Rackspace Cloud Files 0.17.1 using pyftpdlib 0.6.0 ready.
quit
221 Goodbye.
Connection closed by foreign host.
```

EDITED: I posted the wrong log.

Log:

```
Sep 12 13:26:20 juan ftpcloudfs[5398]: Serving FTP on 127.0.0.1:2021
Sep 12 13:26:20 juan ftpcloudfs[5388]: Serving FTP on 127.0.0.1:2021
Sep 12 13:26:20 juan ftpcloudfs[5399]: Serving FTP on 127.0.0.1:2021
Sep 12 13:26:22 juan ftpcloudfs[5399]: []127.0.0.1:43121 Connected.
Sep 12 13:26:22 juan ftpcloudfs[5399]: 127.0.0.1:43121 Connected, shared ip map: {'127.0.0.1': 1}
Sep 12 13:26:22 juan ftpcloudfs[5399]: 127.0.0.1:43121 Connection track: 127.0.0.1 -> 1
Sep 12 13:26:22 juan ftpcloudfs[5399]: 127.0.0.1:43121 ==> 220 Rackspace Cloud Files 0.17.1 using pyftpdlib 0.6.0 ready.
Sep 12 13:26:27 juan ftpcloudfs[5388]: []127.0.0.1:43122 Connected.
Sep 12 13:26:27 juan ftpcloudfs[5388]: 127.0.0.1:43122 Connected, shared ip map: {'127.0.0.1': 2}
Sep 12 13:26:27 juan ftpcloudfs[5388]: 127.0.0.1:43122 Connection track: 127.0.0.1 -> 2
Sep 12 13:26:27 juan ftpcloudfs[5388]: 127.0.0.1:43122 ==> 421 Too many connections from the same IP address.
Sep 12 13:26:27 juan ftpcloudfs[5388]: []@127.0.0.1:43122 Too many connections from the same IP address.
Sep 12 13:26:27 juan ftpcloudfs[5388]: 127.0.0.1:43122 Disconnected, shared ip map: {'127.0.0.1': 1}
Sep 12 13:26:27 juan ftpcloudfs[5388]: []@127.0.0.1:43122 Disconnected.
```

**Comment by reidrac** [comment]
I've moved two log lines that access to the shared dict into the locked code: ac60278bd25693ee212fa2b7bfdab7a2c4bf5a74

I don't know if it's related, but we need to find out why that KeyError happened!

**Comment by sirkonst** [comment]
You made very little attempts, I think. Try about 5 time.

**Comment by sirkonst** [comment]
Thanks for ac60278 and c7b309f.

**Comment by reidrac** [comment]
OK, I see what you mean... I got to reproduce the bug.

Looks like it's not updating the shared dict properly:

```
Sep 12 14:20:50 juan ftpcloudfs[6655]: []127.0.0.1:44340 Connected.
Sep 12 14:20:50 juan ftpcloudfs[6655]: 127.0.0.1:44340 Connected, shared ip map: {'127.0.0.1': 1}
Sep 12 14:20:50 juan ftpcloudfs[6655]: 127.0.0.1:44340 Connection track: 127.0.0.1 -> 1
Sep 12 14:20:50 juan ftpcloudfs[6655]: 127.0.0.1:44340 ==> 220 Rackspace Cloud Files 0.17.1 using pyftpdlib 0.6.0 ready.
Sep 12 14:20:52 juan ftpcloudfs[6655]: []127.0.0.1:44341 Connected.
Sep 12 14:20:52 juan ftpcloudfs[6655]: 127.0.0.1:44341 ==> 421 Too many connections from the same IP address.
Sep 12 14:20:52 juan ftpcloudfs[6655]: []@127.0.0.1:44341 Too many connections from the same IP address.
Sep 12 14:20:52 juan ftpcloudfs[6655]: 127.0.0.1:44341 Disconnected, shared ip map: {}
Sep 12 14:20:52 juan ftpcloudfs[6655]: []@127.0.0.1:44341 Disconnected.
```

See? The second connection ends with an empty dictionary instead of having 1 connection (the first one).

I'm creating an issue for this and we can go on from there.
